llvm/include/llvm/ABI/ABIFunctionInfo.h

```diff
@@ -0,0 +1,172 @@
+//===----- ABIFunctionInfo.h - ABI Function Information ----- C++ ---------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Defines ABIFunctionInfo and associated types used in representing the
+// ABI-coerced types for function arguments and return values.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_ABI_ABIFUNCTIONINFO_H
+#define LLVM_ABI_ABIFUNCTIONINFO_H
+
+#include "ABIInfo.h"
+#include "llvm/ADT/ArrayRef.h"
+#include "llvm/ADT/FoldingSet.h"
+#include "llvm/IR/CallingConv.h"
+#include "llvm/Support/TrailingObjects.h"
+
+namespace llvm {
+namespace abi {
+
+struct FunctionABIInfo {
+  llvm::CallingConv::ID CC = llvm::CallingConv::C;
+  llvm::CallingConv::ID EffectiveCC = llvm::CallingConv::C;
+
+  // Core ABI attributes
+  bool NoReturn = false;
+  bool NoUnwind = false;
```

Comment:
For these properties, we should only include anything that either affects how args/returns are passed or is affected by the args/return.

NoReturn and NoUnwind are optimization attributes, we shouldn't need to track them here.